### PR TITLE
Use on-disk CONIKS trie instead of collections in byzcoin

### PR DESCRIPTION
**Pull request by kc1212**
Key changes:
* Removed collections, changed the backend to use the new library - trie.
* Renamed everything from collectionDB/cdb/collection/coll to stateTrie/st/stagingStateTrie/sst
* `Proof.Match` previously did not take any arguments, which I think is incorrect. If you don't know which key you want to prove, then the service can just send a random inclusion proof for some key that exists. So I changed it to `Proof.Match(key []byte)`.
* The proof structure changed too, of course, which means the JS and Java API needs to be updated. But it's quite easy to check the proof in the new structure, so we can finally implement proper verification in JS/Java.
* Some refactoring here and there.

### CODE REVIEW

**Comment by ineiti** [comment]
Please add links to the top of `byzcoin/trie/README.md`, so that it fits with the rest of the READMEs.

File: byzcoin/contracts.go [hunk 1]
```diff
@@ -302,3 +274,54 @@ func (s *Service) ContractDarc(cdb CollectionView, inst Instruction, coins []Coi
 		return nil, nil, errors.New("unknown instruction type")
 	}
 }
+
+// loadConfigFromTrie loads the configuration data from the trie.
+func loadConfigFromTrie(st ReadOnlyStateTrie) (*ChainConfig, error) {
+	// Find the genesis-darc ID.
+	val, contract, _, err := getValueContract(st, NewInstanceID(nil).Slice())
```
Comment (jeffallen) [inline byzcoin/contracts.go]: Could/should this be using VerifyAndDecode?

File: byzcoin/service.go [hunk 2]
```diff
@@ -1015,31 +1089,27 @@ func (s *Service) createStateChanges(coll *collection.Collection, scID skipchain
 	// we need to find out if this is as expensive as it looks, and if so if
 	// we could use some kind of copy-on-write technique.
 
-	cdbTemp := coll.Clone()
```
Comment (jeffallen) [inline byzcoin/service.go]: The comment above here can go away, since the new data structure was designed for a lightweight clone.

File: external/js/cothority/lib/byzcoin/Proof.js [hunk 3]
```diff
@@ -26,26 +24,22 @@ class Proof {
    * stored, false if it is a proof of absence.
    */
   matches() {
-    return this._leaf !== undefined;
+    // TODO write a proper way to do match
```
Comment (jeffallen) [inline external/js/cothority/lib/byzcoin/Proof.js]: Now or later?

File: external/proto/pop.proto [hunk 4]
```diff
@@ -165,6 +165,7 @@ message GetFinalStatementsReply {
 message StoreInstanceID {
   required bytes partyid = 1;
   required bytes instanceid = 2;
+  required bytes darcid = 3;
```
Comment (jeffallen) [inline external/proto/pop.proto]: Where did these come from? They seem unrelated to this change? (also in GetInstanceIDReply below)

File: byzcoin/proof_test.go [hunk 5]
```diff
@@ -23,36 +23,39 @@ func TestNewProof(t *testing.T) {
 	s := createSC(t)
 	p, err := NewProof(s.c, s.s, skipchain.SkipBlockID{}, []byte{})
 	require.NotNil(t, err)
-	p, err = NewProof(s.c, s.s, s.genesis.Hash, []byte{1})
+
+	key := []byte{1}
+	p, err = NewProof(s.c, s.s, s.genesis.Hash, key)
 	require.Nil(t, err)
-	require.False(t, p.InclusionProof.Match())
+	require.False(t, p.InclusionProof.Match(key))
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/proof_test.go]: see the PR description

File: byzcoin/statetrie.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
+package byzcoin
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/statetrie.go]: do you mean the snapshot feature? my understanding is that it'll be taken care of in a separate struct/function. the statetrie by itself doesn't store history.

[the remark this answers is not in the document]
Comment (ineiti) [reply byzcoin/statetrie.go]: As discussed this morning: adding this functionality would require a big change to `trie`, so currently we'll just use the snapshot-feature. But I think it should go into `statetrie`, as the `trie` doesn't need to be aware of the snapshots, just give the possibility to do them...

File: byzcoin/proof_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ineiti) [reply byzcoin/proof_test.go]: OK - I'm split on this one. I see your point, but it makes using the inclusionproof more difficult. But there is some gain in security by default, because you _have_ to check the key.

So +1 for keeping the key-argument.

File: byzcoin/statetrie.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+// ReadOnlyStateTrie is the read-only interface for StagingStateTrie and
+// StateTrie.
+type ReadOnlyStateTrie interface {
+	GetValues(key []byte) (value []byte, contractID string, darcID darc.ID, err error)
+	GetProof(key []byte) (*trie.Proof, error)
+}
+
+// StagingStateTrie is a wrapper around trie.StagingTrie that allows for use in
+// byzcoin.
+type StagingStateTrie struct {
```
Comment (jeffallen) [inline byzcoin/statetrie.go]: Please consider what's the minimum that can be exported. Maybe just ReadOnlyStateTrie and make StagingStateTrie and StateTrie private so as not to clutter the public API of pkg byzcoin.

File: byzcoin/service.go [hunk 8]
```diff
@@ -887,8 +946,25 @@ func (s *Service) verifySkipBlock(newID []byte, newSB *skipchain.SkipBlock) bool
...
+		var err error
+		cdb, err = NewMemStagingStateTrie(nonce)
+		if err != nil {
+			log.Error(s.ServerIdentity(), err)
+			return false
+		}
+	} else {
+		cdb = s.getStateTrie(newSB.SkipChainID()).MakeStagingStateTrie()
+	}
+	mtr, txOut, scs := s.createStateChanges(cdb.Clone(), newSB.SkipChainID(), body.TxResults, noTimeout)
```
Comment (jeffallen) [inline byzcoin/service.go]: Why is this clone here?

File: byzcoin/service.go [hunk 9]
```diff
@@ -921,14 +997,13 @@ func (s *Service) verifySkipBlock(newID []byte, newSB *skipchain.SkipBlock) bool
 
 	// Compute the new state and check whether the roster in newSB matches
 	// the config.
-	collClone := s.getCollection(newSB.SkipChainID()).coll.Clone()
-	for _, sc := range scs {
-		if err := storeInColl(collClone, &sc); err != nil {
-			log.Error(s.ServerIdentity(), err)
-			return false
-		}
+	eph := cdb.Clone()
```
Comment (jeffallen) [inline byzcoin/service.go]: s/eph/?/

Why Clone here?

File: byzcoin/transaction.go [hunk 10]
```diff
@@ -420,6 +382,45 @@ func (sc StateChange) ShortString() string {
 	return sc.toString(false)
 }
 
+// Key returns the key that should be used in a key/value database.
+func (sc *StateChange) Key() []byte {
```
Comment (jeffallen) [inline byzcoin/transaction.go]: Maybe you can put these methods on StateChange (no *) and then you can remove the copy to []KVPair in StoreAll.

To check the implementation: https://stackoverflow.com/questions/27803654/explanation-of-checking-if-value-implements-interface

File: byzcoin/statetrie.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/statetrie.go]: I see, in that case I think it is possible to extend `StateTrie` to support snapshots (just copy everything to another database/bucket) just by extending the API. You can have an additional field in the `StateTrie` struct that stores a pointer to the snapshot database. Then have a new function that executes the `CopyTo` function of `Trie` whenever a snapshot is needed. Perhaps we create a new bucket for every snapshot, maybe in a new database, the bucket name can be derived from skipblock ID so that we can quickly find what we're looking for or iterate through all the buckets to see what snapshots we have. The nested bucket feature - https://github.com/etcd-io/bbolt#nested-buckets might be useful here because we might store all the snapshots in one bucket but each snapshot in their own buckets.

File: byzcoin/service.go [hunk 11]
```diff
@@ -186,13 +180,17 @@ func (s *Service) CreateGenesisBlock(req *CreateGenesisBlock) (
 		return nil, err
 	}
 
+	// This is the nonce for the trie.
+	nonce := GenNonce()
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: i've changed it to private, it won't be needed anyway after https://github.com/dedis/cothority/issues/1442 is done because it won't use a nonce anymore

File: byzcoin/service.go [hunk 12]
```diff
@@ -413,7 +409,14 @@ func (s *Service) createNewBlock(scID skipchain.SkipBlockID, r *onet.Roster, tx
 		// We have to register the verification functions in the genesis block
 		sb.VerifierIDs = []skipchain.VerifierID{skipchain.VerifyBase, verifyByzCoin}
 
-		coll = collection.New(&collection.Data{}, &collection.Data{}, &collection.Data{})
+		// Staging tries are only allowed to be created from concrete
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: not sure what I was writing, probably a comment for something I removed...

File: byzcoin/service.go [hunk 13]
```diff
@@ -526,23 +548,23 @@ func (s *Service) updateCollectionCallback(sbID skipchain.SkipBlockID) error {
 		return errors.New("couldn't unmarshal body")
 	}
 
-	log.Lvlf2("%s Updating transactions for %x", s.ServerIdentity(), sb.SkipChainID())
-	_, _, scs := s.createStateChanges(cdb.coll, sb.SkipChainID(), body.TxResults, noTimeout)
+	log.Lvlf2("%s Updating transactions for %x on index %d", s.ServerIdentity(), sb.SkipChainID(), sb.Index)
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: no idea if it matters, maybe it'll give you a compile error if you accidentally use some other type for `%d`

File: byzcoin/service.go [hunk 14]
```diff
@@ -653,26 +677,56 @@ func isViewChangeTx(txs TxResults) *viewchange.View {
...
+	if s.stateTries[idStr] != nil {
+		// Usually this function shouldn't be called if the trie
+		// already exists, so we return early.
+		return nil
+	}
+	db, name := s.GetAdditionalBucket([]byte(idStr))
+	st, err := NewStateTrie(db, name, nonce)
+	if err != nil {
+		log.Error(s.ServerIdentity(), idStr, err)
+		return nil
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: in this case, the `nil` means an error, i.e., we failed to load/create the state trie. i kept it this way because the old `getCollections` was also doing it this way

File: byzcoin/statetrie.go [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+	}
+	return &StateTrie{
+		Trie: *t,
+	}, nil
+}
+
+// NewStateTrie creates a new, disk-based trie.Trie, an error is returned if
+// the db already contains a trie.
+func NewStateTrie(db *bolt.DB, bucket, nonce []byte) (*StateTrie, error) {
+	t, err := trie.NewTrie(trie.NewDiskDB(db, bucket), nonce)
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/statetrie.go]: I don't quite understand the first question. Why does `NewTrie` need to use `metadata.go`? If you mean it might be better to use the `metadata.go` functions to load the nonce, then I think it's not possible because we haven't created the trie yet.

Nice catch regarding `db.Update` in `LoadTrie`.

File: byzcoin/service.go [hunk 16]
```diff
@@ -504,11 +506,31 @@ func (s *Service) updateCollectionCallback(sbID skipchain.SkipBlockID) error {
...
+		err := protobuf.DecodeWithConstructors(sb.Payload, &body, network.DefaultConstructors(cothority.Suite))
+		if err != nil {
+			log.Error(s.ServerIdentity(), "could not unmarshal body for genesis block", err)
+			return errors.New("couldn't unmarshal body for genesis block")
+		}
+		nonce := body.TxResults[0].ClientTransaction.Instructions[0].Spawn.Args.Search("trie_nonce")
+		if len(nonce) == 0 {
+			return errors.New("nonce is empty")
+		}
+		s.createStateTrie(sb.SkipChainID(), nonce)
```
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: this was intentional actually, because afterwards we use `getStateTrie`. if something goes wrong then an error is returned. i added a comment to explain why the returned value isn't used

File: byzcoin/contracts.go (hunk 1 above)
Comment (kc1212) [reply byzcoin/contracts.go]: i'm not sure why this would be `VerifyAndDecode`, which is a method of `Proof`. these functions are copy pasted from `struct.go` because they're mostly (only?) used for the contracts.

File: external/js/cothority/lib/byzcoin/Proof.js (hunk 3 above)
Comment (kc1212) [reply external/js/cothority/lib/byzcoin/Proof.js]: in a future PR, because java/js never had a proper `match` implementation

File: external/proto/pop.proto (hunk 4 above)
Comment (kc1212) [reply external/proto/pop.proto]: It's the result of re-running `proto.sh`. I think one of the earlier PRs changed a `proto.go` file but didn't re-run `proto.sh`.

File: byzcoin/transaction.go (hunk 10 above)
Comment (kc1212) [reply byzcoin/transaction.go]: your suggestion didn't work, I made a simple test here - https://play.golang.org/p/waJ1f-6pz8a

regardless of whether it's a pointer receiver or a regular receiver, the `func addAreas(shapes []Shape) int` won't compile...

File: byzcoin/service.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (kc1212) [reply byzcoin/service.go]: actually, I had to bring this back because the simulation test fails if we don't use real nonces. but anyway, the `GenNonce` will go away when we have proper client transactions

**Comment by kc1212** [comment]
Addressed comments

File: byzcoin/service.go [hunk 17]
```diff
@@ -1245,6 +1242,21 @@ func (s *Service) getTxs(leader *network.ServerIdentity, roster *onet.Roster, sc
 	return s.txBuffer.take(string(scID))
 }
 
+func (s *Service) loadNonceFromTxs(txs TxResults) ([]byte, error) {
+	if len(txs) == 0 {
+		return nil, errors.New("no transactions")
+	}
+	instrs := txs[0].ClientTransaction.Instructions
```
Comment (jeffallen) [inline byzcoin/service.go]: Two more checks to be even more conservative: if len(instrs) != 1 { return nil, fmt.Errorf("expected 1 instruction, got %v", len(instrs)) } and if instrs[0].Spawn == nil { return nil, errors.New("first instruction is not a Spawn") }

File: byzcoin/statetrie.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jeffallen) [reply byzcoin/statetrie.go]: Never mind about the metadata question, I figured out how getIndex works later.
